### ISSUE TITLE
New lint to not use `{}` in generic

### ISSUE THREAD

**Issue by donyunardi** [jimhester/lintr#1968]
Based on Hadley's suggestion:
https://github.com/hadley/adv-r/blob/dc49c3872c3530ac08716fd4f4c235b01266a4ce/S4.Rmd#L345-L352

> It is bad practice to use `{}` in the generic as it triggers a special case that is more expensive, and generally best avoided.
>
> ```{r}
> # Don't do this!
>  setGeneric("myGeneric", function(x) {
>  standardGeneric("myGeneric")
> })
> ```

Would it be a good idea to create a lint for this rule?

Linking with:
https://github.com/insightsengineering/teal.code/issues/109

**Comment by MichaelChirico** [comment]
SGTM. It will be part of `brace_linter()`. Note that there are two things:

 - Throw a lint if `{` is used in the `def=` (2nd positional) argument of `setGeneric`, either `\(...) { ... }` or `function(...) { ... }`
 - _Don't_ throw a lint if `{` is excluded from a multi-line function in the `def=` argument of `setGeneric`

The latter can come up for long definitions like

```r
setGeneric(
  "myLongGeneric",
  function(a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p)
    standardGeneric("myLongGeneric")
)
```

where bumping the line is used to avoid `line_length_linter()` hits.

**Comment by MichaelChirico** [comment]
Hmm actually in the multi-line case, styler will auto-include the `{` as well. So might need a companion issue, WDYT @lorenzwalthert?

**Comment by MichaelChirico** [comment]
Linking {styler} tracker: https://github.com/r-lib/styler/issues/1141

**Comment by Bisaloo** [comment]
I had a look at this and I don't think it's possible to do it while having the `linter_level` set to expression.

Indeed, this we need to match the `setGeneric()` first argument to the `standardGeneric()` argument, we need to use information across multiple consecutive expressions.

Am I missing something?

**Comment by MichaelChirico** [comment]
I deployed a friendly neighborhood LLM (Gemini) to investigate this in a bit more detail.

My own takeaway is that yes, it would probably be best for R itself to just strip the `{` and give `standardGeneric()`.

----

# Analysis of S4 Generic definition with and without `{}`

This report investigates the claim that using curly braces `{}` in S4 generics in R is bad practice because it triggers a more expensive special case.

## Summary of Findings

1.  **There is a structural difference**:
    *   Generics defined **without** `{}` (e.g., `setGeneric("f", function(x) standardGeneric("f"))`) are assigned the class **`standardGeneric`**.
    *   Generics defined **with** `{}` (e.g., `setGeneric("f", function(x) { standardGeneric("f") })`) are assigned the class **`nonstandardGenericFunction`**.
    *   The body of `standardGeneric` is normalized to `standardGeneric("name")`. The body of `nonstandardGenericFunction` retains the curly braces.

2.  **There is a performance difference in uncompiled code**:
    *   In interpreted (uncompiled) R code, the curly braces add a small but measurable overhead due to the AST interpreter having to evaluate the `{` call.
    *   In our benchmarks, uncompiled S4 dispatch was **3% to 6% slower** when using curly braces.

3.  **There is NO performance difference in compiled code**:
    *   The R byte-code compiler optimizes away single-expression curly braces.
    *   When compiled, both forms perform identically.
    *   Since R packages are byte-compiled by default during installation, **there is no performance difference for generics defined in packages**.

---

## Detailed Analysis

### 1. Structural Difference: `standardGeneric` vs `nonstandardGenericFunction`

When `setGeneric` is called, it tests if the supplied definition is "standard" using the internal function `.NonstandardGenericTest`.

If the definition body is identical to `standardGeneric("name")` (without curly braces), it is considered standard.
If the body is wrapped in curly braces, like `{ standardGeneric("name") }`, it fails the identity test and is classified as `nonstandardCase = TRUE`.

This classification leads to two main differences:
*   **Class**: Standard generics get class `standardGeneric` (which inherits from `genericFunction`). Non-standard generics get class `nonstandardGenericFunction` (which inherits from `genericFunction` and `nonstandardGeneric`).
*   **Normalization**: For standard generics, the body is replaced by a clean `standardGeneric("name")` call. For non-standard generics, the user-supplied body (with curly braces) is kept.

### 2. Performance Quantification

We ran benchmarks comparing the two definition styles under different compilation settings.

#### Benchmark Setup

We defined class `A` and an object `x` of class `A`. We defined two generics `f1` (standard) and `f2` (non-standard) with corresponding methods:

```R
setClass("A", representation(x = "numeric"))
x <- new("A", x = 1)

# Standard (no curly braces)
setGeneric("f1", function(obj) standardGeneric("f1"))
setMethod("f1", "A", function(obj) obj@x)

# Non-standard (with curly braces)
setGeneric("f2", function(obj) { standardGeneric("f2") })
setMethod("f2", "A", function(obj) obj@x)
```

We measured the user time for 1,000,000 calls to `f1(x)` and `f2(x)`.

#### Results

| Scenario | `f1` (Standard, no `{}`) | `f2` (Non-standard, with `{}`) | Ratio (`f2`/`f1`) |
| :--- | :--- | :--- | :--- |
| **Uncompiled (JIT 0)** | 1.416s | 1.477s | **1.043** (~4% slower) |
| **Compiled (explicit)** | 2.695s* | 2.704s* | **1.003** (identical) |

*\*Note: Compiled benchmark was run with 2,000,000 iterations, hence the higher absolute times. The ratio is the key metric.*

#### Interpretation

*   **Uncompiled**: The ~4% overhead in the uncompiled case is due to the interpreter evaluating the `{` call. This overhead is relatively small because S4 dispatch itself is heavy, but it is consistent.
*   **Compiled**: The R compiler disassembles both functions to nearly identical bytecode, effectively removing the `{` wrapper.

### Bytecode Comparison

Disassembling the compiled versions of the two generics shows that both call `standardGeneric` directly:

**`f1` (Standard) Bytecode:**
```
GETFUN.OP (standardGeneric)
PUSHCONSTARG.OP ("f1")
CALL.OP
RETURN.OP
```

**`f2` (Non-standard) Bytecode:**
```
GETFUN.OP (standardGeneric)
PUSHCONSTARG.OP ("f2")
CALL.OP
RETURN.OP
```

The only difference in the compiled objects is the metadata (constants list) which retains the original source expression for debug/source reference purposes, but this does not affect execution speed.

## Implications of Class Difference

The class difference (`standardGeneric` vs `nonstandardGenericFunction`) has minor implications:

1.  **Inconsistent Error Messages**: In `.findInheritedMethods` (an internal helper in `methods` package), if the methods table is corrupted or invalid, R checks the class of the generic to report an error:
    *   If it is `standardGeneric`, it reports: `invalid or unset methods table in generic function '...'`.
    *   Otherwise, it reports: `trying to find a methods table in a non-generic function`.
    Because `nonstandardGenericFunction` does not inherit from `standardGeneric`, it gets the second, misleading error message, even though it IS a generic function.

2.  **Potential Third-Party Dependency Issues**: Third-party packages that inspect S4 generics might incorrectly assume all simple generics are of class `standardGeneric` instead of checking inheritance from `genericFunction` or using `isGeneric()`. However, standard R APIs handle this correctly.

## Is it a bug in R?

It can be considered a minor design inconsistency or oversight:

1.  **Strict Syntax Check**: R's `.NonstandardGenericTest` uses `identical(body, stdBody)` which is a strict syntactic check. It does not recognize that `{ standardGeneric("f") }` is semantically identical to `standardGeneric("f")`.
2.  **Inaccurate Error Check**: The check in `.findInheritedMethods` should check for `genericFunction` class rather than `standardGeneric` specifically, to avoid misleading error messages for non-standard generics.

A potential fix in R would be to:
*   Allow `.NonstandardGenericTest` to recognize single-expression curly braces as standard.
*   Normalize the body by removing the curly braces if it is classified as standard.
*   Update internal checks in `methods` package to look for `genericFunction` instead of `standardGeneric` when checking if an object is a generic.

## Conclusion

The claim in the issue is **technically true for uncompiled code**, but **practically irrelevant for compiled package code** in terms of performance.

However, avoiding curly braces in S4 generics remains a best practice because:
1.  It ensures the generic is classified as `standardGeneric`, avoiding misleading error messages in case of internal corruption.
2.  It avoids any interpreter overhead if the code is run uncompiled (e.g., during development or in scripts with JIT disabled).
3.  It results in cleaner code that matches the documentation recommendations.

**Comment by MichaelChirico** [comment]
I filed https://bugs.r-project.org/show_bug.cgi?id=19083

**Comment by Bisaloo** [comment]
I've also been investigating this on and off during the R contributor office hours and came to a similar conclusion. It seems like an historical quirk that braced standard generic calls are identified as non-standard.

Sharing the WIP write up I had about this:

------

# Compare S4 standardGeneric() special case to non-special case


S4 style guides usually recommend not adding braces around `standardGeneric()` function bodies. For examples, the [‘S4’ chapter of ‘Advanced R’](https://adv-r.hadley.nz/s4.html#s4-generics) states:

> It is bad practice to use {} in the generic as it triggers a special case that is more expensive, and generally best avoided.
>
> ``` r
> # Don't do this!
> setGeneric("myGeneric", function(x) {
>   standardGeneric("myGeneric")
> })
> ```

This is a common source of confusion for novice S4 users, and for tools analyzing or modifying R code formatting (https://github.com/posit-dev/air/issues/398, https://github.com/r-lib/lintr/issues/1968, https://github.com/hadley/adv-r/issues/1804) Indeed, it is unusual in R to have formatting choices result in different behaviours. Furthermore, the exact consequence of the special case seem to have been lost in time and remains unknown to many R users, including experienced S4 developers. It is unclear if the recommendation is still necessary today, or if we are just following a historical convention.

This blog post explores where this special case arises from, and what its measurable consequences (if any) are.

## Special case definition

The source of the special case was pointed out to me by Lionel Henry on the Air bug tracker. [In the methods package, there is an extra check that will compare the body of our defined generic to a known pattern for `standardGeneric()` calls without braces](https://github.com/r-devel/r-svn/blob/097aac926efd9beed35fec21b4f2be14ad8453b4/src/library/methods/R/RMethodUtils.R#L1227-L1235).

> ``` r
> .NonstandardGenericTest <- function(body, fname, stdBody)
> {
>     if(identical(body, stdBody))
>         FALSE
>     else if(.recursiveCallTest(body, fname))
>         TRUE
>     else
>         NA
> }
> ```

As far as I can tell, the initial version of this “nonStandardGenericFunction” was introduced in [revision 20995](https://github.com/r-devel/r-svn/commit/62c60532209aca868d5ec4560abcb49e53004f9d).

## Special case consequences

To see the consequences of triggering the special case or not, let’s define two generics, which do the same thing, but one triggers the special case and the other does not:

``` r
# This triggers the \pkg{methods} special case
setGeneric("myStandardGeneric", function(x) standardGeneric("myStandardGeneric"))
#> [1] "myStandardGeneric"

# This doesn't trigger the \pkg{methods} special case because of the braces
setGeneric("myCustomGeneric", function(x) {
  standardGeneric("myCustomGeneric")
})
#> [1] "myCustomGeneric"
```

And then let’s define a simple method for both generics:

``` r
setOldClass("rle")

setMethod("myStandardGeneric", "rle", function(x) {
  x$values * x$lengths / sum(x$lengths)
})

setMethod("myCustomGeneric", "rle", function(x) {
  x$values * x$lengths / sum(x$lengths)
})
```

### Performance

[What the methods package documentation seems to be suggesting](https://github.com/r-devel/r-svn/blob/56b976ae0de9cec972c8a90a8d301a9866094869/src/library/methods/man/Methods_Details.Rd#L282-L287), and what is often repeated, is that the standard generic shortcut may allow for better performance. We can try it on our simple example:

``` r
x <- rpois(1000, lambda = 3) |> rle()
microbenchmark::microbenchmark(
  myStandardGeneric(x),
  myCustomGeneric(x),
  times = 1e6
)
#> Unit: microseconds
#>                  expr   min    lq     mean median    uq      max neval cld
#>  myStandardGeneric(x) 4.741 5.153 5.919138  5.301 5.503 36212.91 1e+06   a
#>    myCustomGeneric(x) 4.803 5.188 5.908835  5.338 5.542 34432.95 1e+06   a
```

This doesn’t seem to have any measurable impact on performance.

On the topic of performance, [an email in the R-devel mailing list](https://www.mail-archive.com/r-devel@r-project.org/msg06307.html) says:

> standardGeneric() creates a new context into which it copies bindings for all the objects in the frame of the generic function call

But my reading of this is that objects created before calling `standardGeneric()` in a non-standard generic that does some initial pre-computation would be copied into the context. It doesn’t apply in the simple case of a non-standard generic that just wraps `standardGeneric()` in braces, as it doesn’t do any pre-computation.

### Error stack

One GitHub comment mentions “an additional function call”. If this the case, even if it doesn’t have a measurable impact on performance, it could make the error stack more difficult to understand, and make debugging more difficult.

Let’s try to trigger an error in both cases:

``` r
broken_rle <- structure(list(lengths = letters, values = 1:26), class = "rle")
myStandardGeneric(broken_rle)
#> Error in `x$values * x$lengths`:
#> ! non-numeric argument to binary operator
myCustomGeneric(broken_rle)
#> Error in `x$values * x$lengths`:
#> ! non-numeric argument to binary operator
```

The error stack appears to be the same in both cases, and it doesn’t seem that the special case adds an extra function call.

### Generic collision across packages

(Double check. But might a mechanism to avoid generic collision across packages be the original motivation for this special case? ?setGeneric() talks a bit about this in the “Specialized Local Generics” section.)

### Other issues

There might be other consequences of triggering the special case that I did not identify. Some active Bioconductor package developers could not recall precisely **why** one should try to target the special case.

I did a [search on the Bioconductor mailing list](https://www.mail-archive.com/search?l=bioc-devel%40r-project.org&q=standardGeneric&start=100) and on the \[Bioconductor slack/zulip\] with the text “standardGeneric” and could not find any relevant threads.

## Summary

There doesn’t seem to be any practical impact of using a braced `standardGeneric()` call, whether in terms of performance, error stack, or anything else that I could identify.

However, there seem to be a lot of lost knowledge around the original motivation for this special case, and it is possible that there are consequences that I did not identify. I would love for this blog post to be a starting point for a discussion on this topic, and for people to share any information they might have on the original motivation for this special case, and any consequences that I might have missed.

## Recommendations

### Advanced R book

It would be helpful to explain exactly **why** one should try to trigger the special case. If this is not appropriate for the main text, perhaps an aside or footnote could be added.

The current phrasing (“more expensive”) leads to believe the main issue is degraded performance, when this blog post seems to point in another direction.

### methods package

Upon exploring this issue, it appeared that non standard generics were originally meant to identify cases where some pre-computation is done in the generic rather than the methods, [which is discouraged](https://www.mail-archive.com/r-devel@r-project.org/msg06307.html):

> Maybe more relevant, if this question is motivated by a practical need, you may want to rethink what the generic function does. “Nonstandard” generic functions are called that for a reason. The concept behind methods in S is that a generic function defines the *function* or purpose of the computation, but it’s only the methods that define how the purpose is carried out.
>
> Using a nonstandard generic that does some extra computation always blurs the concept, and often makes methods harder to write or understand.

But braced `standardGeneric()` calls do not actually perform any pre-computation, and it seems that it is purely a technical and historical quirk that they are treated as non-standard generics.

It would be helpful if braced `standardGeneric()` calls could be identified as standard generics.

This would align better with the rest of the R ecosystem, where formatting choices usually don’t have an impact on the behaviour of the code, and it would also make it easier for tools analyzing or modifying R code formatting to handle `standardGeneric()` calls without having to worry about the presence of braces.

There are several ways this could be achieved:

- By unbracing the generic body before the check in `.NonstandardGenericTest()`
- By comparing the body to both the braced and unbraced versions of the `standardGeneric()` call
